bittide/src/Bittide/Transceiver/Prbs.hs

```diff
@@ -0,0 +1,137 @@
+-- SPDX-FileCopyrightText: 2023-2024 Google LLC
+--
+-- SPDX-License-Identifier: Apache-2.0
+
+{-# LANGUAGE GADTs #-}
+
+-- | A pseudo-random bit sequence (PRBS) generator and checker. These functions
+-- are used to test the signal integrity of transceivers. The generator generates
+-- a PRBS stream, while the checker checks whether the received stream is the
+-- same as the generated stream. Note that the checker is "self synchronizing",
+-- meaning that it will synchronize with the generator after /polyLength/ cycles.
+module Bittide.Transceiver.Prbs where
+
+import Clash.Explicit.Prelude
+
+-- | Configuration for a PRBS generator or checker. Note that this can only
+-- specify PRBS streams with two taps: the first tap is always the MSB
+-- (@polyLength@), and the second tap is the @polyTap@-th bit.
+--
+-- See https://en.wikipedia.org/wiki/Pseudorandom_binary_sequence for more information.
+data Config polyLength polyTap nBits where
+  Config ::
+    ( KnownNat polyLength
+    , KnownNat polyTap
+    , KnownNat nBits
+
+    , 1 <= nBits
+    , 1 <= polyTap
+    , (polyTap + 1) <= polyLength
+
+    -- Same constraints, but written differently for type checking purposes:
+    , (_n0 + 1) ~ nBits
+    , (polyTap + _n1) ~ polyLength
+    , polyTap ~ (_n2 + 1)
+    , _n1 ~ (_n3 + 1)
+    ) =>
+    Config polyLength polyTap nBits
+
+
+-- | PRBS31: @x^31 + x^28 + 1@
+conf31 :: forall n . (31 <= n, KnownNat n) => Config 31 28 n
+conf31 = leToPlus @31 @n Config
+
+-- | PRBS generator, see module documentation.
+generator ::
+  forall dom polyLength polyTap nBits .
+  KnownDomain dom =>
+  Clock dom -> Reset dom -> Enable dom ->
+  Config polyLength polyTap nBits ->
+  Signal dom (BitVector nBits)
+generator clk rst ena Config =
+  moore clk rst ena go snd (maxBound, maxBound) (pure ())
+ where
+  go ::
+    (BitVector polyLength, BitVector nBits) ->
+    () ->
+    (BitVector polyLength, BitVector nBits)
+  go (prbs_reg, _) _ =
+    ( last prbs
+    , pack (reverse $ map msb prbs) )
+   where
+     prbs :: Vec nBits (BitVector polyLength)
+     prbs = unfoldrI goPrbs prbs_reg
+
+     goPrbs :: BitVector polyLength -> (BitVector polyLength, BitVector polyLength)
+     goPrbs bv = (o,o)
+      where
+       o = newBit +>>. bv
+       tap = SNat @(polyLength - polyTap)
+       newBit = xor (lsb bv) (unpack $ slice tap tap bv)
```

Comment:
Slicing a single bit looks a bit cumbersome, why not
```haskell
xor (lsb bv) $ boolToBit $ testBit bv $ natToNum @(polyLength - polyTap)
```

Reply:
I did this, but it broke the transceivers :melting_face:. I don't really have the time to debug this, so if you don't mind I'll let it be.

Reply:
Can you document this in a comment then, so we don't mess up with it any time in the future soon?